sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,187 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/freestanding/limits.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aon_timer.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_rv_timer.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aon_timer_testutils.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rand_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+// TODO: [arch,aon] Create a set of target specific AON clock frequency
+// constants #9927
+const uint64_t kClockFreqAonHz = 125 * 1000;  // 125kHz
+
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static const uint32_t kTickFreqHz = 1000 * 1000;  // 1Mhz / 1us
+static dif_aon_timer_t aon_timer;
+static dif_rv_timer_t rv_timer;
+static dif_rv_plic_t plic;
+
+static volatile dif_aon_timer_irq_t irq;
+static volatile top_earlgrey_plic_peripheral_t peripheral;
+static volatile uint64_t time_elapsed;
+
+/**
+ * Initialize the rv timer to count the tick.
+ *
+ * The `ibex_mcycle_read` function uses the `mcycle` register to count
+ * instructions cycles executed and measure time elapsed between two instants,
+ * however it stops counting when the `wfi` is called. As this test is based on
+ * the `wfi` instruction the best approach then to measure the time elapsed is
+ * to use the mtime register, which is basically attached to rv_timer in the
+ * opentitan.
+ * https://docs.opentitan.org/hw/ip/rv_timer/doc/
+ *
+ * This is fine due to the test running in a single thread of execution,
+ * however, the care should be taken in case it changes. OTTF configures the
+ * timer in vPortSetupTimerInterrupt, and re-initialising it inside the test
+ * could potentially break or cause unexpected behaviour of the test framework.
```

Comment:
This only happens if the test is run as a FreeRTOS task, i.e., the `enable_concurrency` flag is set in the `kTestConfig` struct, (instead of bare-metal, which is the default), and if `configUSE_PREEMPTION` is set in `sw/device/lib/testing/test_framework/FreeRTOSConfig.h`. Maybe you could add a `static_assert` here to check the configuration to warn if this test  compiled for a non-supported configuration?